### PR TITLE
Refactor focusing to contextLine

### LINKED ISSUE #24613: [Bug]: Cursor jumps after first edit
### What happened?

https://user-images.githubusercontent.com/44893228/206675556-804aa69b-341c-4e13-ba65-f45a165b87b3.mp4

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [X] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by alexeyinkin**
- Resolves #24613 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart [hunk 1]
```diff
@@ -16,31 +16,120 @@
...
+        child: CodeTheme(
+          data: ext.codeTheme,
+          child: Container(
+            color: ext.codeTheme.styles['root']?.backgroundColor,
+            child: SingleChildScrollView(
+              controller: _scrollController,
+              child: CodeField(
+                key: ValueKey(widget.controller.codeController),
+                focusNode: _focusNode,
+                enabled: isEnabled,
```
Comment (nausharipov) [inline playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart:110]: Sort?

**Review by nausharipov — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart [hunk 2]
```diff
@@ -16,31 +16,120 @@
...
+      textField: true,
+      multiline: true,
+      enabled: isEnabled,
+      readOnly: isEnabled,
+      label: 'widgets.codeEditor.label',
+      child: FocusScope(
+        node: FocusScopeNode(canRequestFocus: isEnabled),
+        child: CodeTheme(
+          data: ext.codeTheme,
+          child: Container(
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart:103]: It requires non-nullable color, ours is nullable.

File: playground/frontend/playground_components/lib/src/controllers/snippet_editing_controller.dart [hunk 3]
```diff
@@ -100,6 +103,27 @@ class SnippetEditingController extends ChangeNotifier {
...
+
+    if (contextLine1Based == null) {
+      return;
+    }
+
+    _toStartOfFullLine(max(contextLine1Based - 1, 0));
+  }
+
+  void _toStartOfFullLine(int line) {
+    final fullPosition = codeController.code.lines.lines[line].textRange.start;
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/controllers/snippet_editing_controller.dart:121]: Added a check.

File: playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart [hunk 4]
```diff
@@ -16,31 +16,120 @@
...
   Widget build(BuildContext context) {
-    return EditorTextArea(
-      codeController: controller.codeController,
-      sdk: controller.sdk,
-      enabled: !(controller.selectedExample?.isMultiFile ?? false),
-      example: controller.selectedExample,
-      isEditable: isEditable,
-      goToContextLine: goToContextLine,
+    final ext = Theme.of(context).extension<BeamThemeExtension>()!;
+    final isMultifile = widget.controller.selectedExample?.isMultiFile ?? false;
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart]: Capitalized.

File: playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart [hunk 5]
```diff
@@ -16,31 +16,120 @@
...
+    final isEnabled = widget.isEditable && !isMultifile;
+
+    return Semantics(
+      container: true,
+      textField: true,
+      multiline: true,
+      enabled: isEnabled,
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart:94]: `+`

File: playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart (hunk 1 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/widgets/snippet_editor.dart]: `+`

File: playground/frontend/playground_components/lib/src/controllers/snippet_editing_controller.dart [hunk 6]
```diff
@@ -100,6 +103,27 @@ class SnippetEditingController extends ChangeNotifier {
     }
   }
 
+  void _toStartOfContextLineIfAny() {
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/controllers/snippet_editing_controller.dart:106]: I would rather keep this.

**Comment by alexeyinkin** [comment]
R: @damondouglas

**Review by nausharipov — APPROVED** [review]

**Review by damondouglas — APPROVED** [review]
@pabloem LGTM

**Comment by pabloem** [comment]
lgtm thanks

**Pull request merged**